src/cache.c

```diff
@@ -0,0 +1,388 @@
+{
+    head->next = head;
+    head->prev = head;
+}
+
+static inline void list_add(struct list_head *node, struct list_head *head)
+{
+    struct list_head *next = head->next;
+
+    next->prev = node;
+    node->next = next;
+    node->prev = head;
+    head->next = node;
+}
+
+static inline void list_del(struct list_head *node)
+{
+    struct list_head *next = node->next;
+    struct list_head *prev = node->prev;
+
+    next->prev = prev;
+    prev->next = next;
+}
+
+static inline void list_del_init(struct list_head *node)
+{
+    list_del(node);
+    INIT_LIST_HEAD(node);
+}
+
+#define list_entry(node, type, member) container_of(node, type, member)
+
+#define list_last_entry(head, type, member) \
+    list_entry((head)->prev, type, member)
+
+#ifdef __HAVE_TYPEOF
+#define list_for_each_entry_safe(entry, safe, head, member)                \
+    for (entry = list_entry((head)->next, __typeof__(*entry), member),     \
+        safe = list_entry(entry->member.next, __typeof__(*entry), member); \
+         &entry->member != (head); entry = safe,                           \
+        safe = list_entry(safe->member.next, __typeof__(*entry), member))
+#else
+#define list_for_each_entry_safe(entry, safe, head, member, type) \
+    for (entry = list_entry((head)->next, type, member),          \
+        safe = list_entry(entry->member.next, type, member);      \
+         &entry->member != (head);                                \
+         entry = safe, safe = list_entry(safe->member.next, type, member))
+#endif
+
+#define INIT_HLIST_HEAD(ptr) ((ptr)->first = NULL)
+
+static inline void INIT_HLIST_NODE(struct hlist_node *h)
+{
+    h->next = NULL;
+    h->pprev = NULL;
+}
+
+static inline int hlist_empty(const struct hlist_head *h)
+{
+    return !h->first;
+}
+
+static inline void hlist_add_head(struct hlist_node *n, struct hlist_head *h)
+{
+    struct hlist_node *first = h->first;
+    n->next = first;
+    if (first)
+        first->pprev = &n->next;
+
+    h->first = n;
+    n->pprev = &h->first;
+}
+
+static inline int hlist_unhashed(const struct hlist_node *h)
+{
+    return !h->pprev;
+}
+
+static inline void hlist_del(struct hlist_node *n)
+{
+    struct hlist_node *next = n->next;
+    struct hlist_node **pprev = n->pprev;
+
+    *pprev = next;
+    if (next)
+        next->pprev = pprev;
+}
+
+static inline void hlist_del_init(struct hlist_node *n)
+{
+    if (hlist_unhashed(n))
+        return;
+    hlist_del(n);
+    INIT_HLIST_NODE(n);
+}
+
+#define hlist_entry(ptr, type, member) container_of(ptr, type, member)
+
+#ifdef __HAVE_TYPEOF
+#define hlist_entry_safe(ptr, type, member)                  \
+    ({                                                       \
+        typeof(ptr) ____ptr = (ptr);                         \
+        ____ptr ? hlist_entry(____ptr, type, member) : NULL; \
+    })
+#else
+#define hlist_entry_safe(ptr, type, member) \
+    ({ (ptr) ? hlist_entry(ptr, type, member) : NULL; })
```

Comment:
The pair `({` and `})` is indeed a GNU extension, and we shall avoid using it for non-gcc/clang compilers. Later, rv32emu might support MSVC. Thus, we would write portable code as well.